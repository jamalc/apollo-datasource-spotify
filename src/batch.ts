import { PagingObject } from '../types/generated';

interface Params {
  limit?: number;
  offset?: number;
}

interface Options<D> {
  /* Size of each batch call */
  size?: number;
  waitForCallback?: boolean;
  /* Maximum number of items to fetch */
  limit?: number;
  offset?: number;
  request: (params: Params) => Promise<PagingObject & { items: D[] }>;
  callback: (data: PagingObject & { items: D[] }) => Promise<void>;
}

export const batch = async <D>({
  size = 50,
  waitForCallback = true,
  limit = 50,
  offset = 0,
  request,
  callback,
}: Options<D>): Promise<void> => {
  let pending = limit;
  let params = {
    offset: offset,
    limit: Math.min(size, pending),
  };

  do {
    const data = await request(params);
    const cb = callback(data);
    if (waitForCallback) {
      await cb;
    }

    pending = Math.min(data.total, pending) - size;
    params = {
      offset: params.offset + size,
      limit: Math.min(size, pending),
    };
  } while (pending > 0);
};
