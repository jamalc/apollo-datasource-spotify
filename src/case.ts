import _ from 'lodash';
import _fp from 'lodash/fp';

interface Converter {
  (key: string): string;
}

interface WalkFn {
  (node: any): any;
}

function createIteratee(converter: Converter, walk: WalkFn) {
  return (result: any, value: any, key: string) =>
    _.set(
      result,
      converter(key),
      _fp.isObjectLike(value) ? walk(value) : value
    );
}

function createTransformer(keyConverter: Converter) {
  return function walk(node: any): any {
    if (_.isArray(node)) return _fp.map(walk, node);
    if (_fp.isPlainObject(node))
      return _.transform(node, createIteratee(keyConverter, walk));
    return node;
  };
}

export function camelCase(o: any) {
  return createTransformer(_fp.camelCase)(o);
}

export function snakeCase(o: any) {
  return createTransformer(_fp.snakeCase)(o);
}
