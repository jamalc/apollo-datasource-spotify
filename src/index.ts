export { SpotifyAPI, Context } from './api';
export { batch } from './batch';
export * from './objects';
export * from './resolvers';
export * from './responses';
export * from './requests';
export * from './typeDefs';
