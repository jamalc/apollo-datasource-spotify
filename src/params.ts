export interface Params {
  [key: string]: any;
}

export interface Page {
  limit?: number;
  offset?: number;
}

export interface Cursor {
  limit?: number;
  after?: number;
  before?: number;
}

export interface AlbumTracks extends Params, Page {}

export interface ArtistAlbums extends Params, Page {
  include_groups?: 'album' | 'single' | 'appears_on' | 'compilation';
}

export interface ArtistTopTracks extends Params {
  country: string;
}

export interface Saved extends Params, Page {}

export interface Top extends Params, Page {
  time_range?: 'short_term' | 'medium_term' | 'long_term';
}

export interface AddTracksToPlaylist extends Params {
  uris?: string[];
  position?: number;
}

export interface UpdatePlaylist extends Params {
  name?: string;
  public?: boolean;
  collaborative?: boolean;
  description?: string;
}

export interface CreatePlaylist extends Params {
  name?: string;
  public?: boolean;
  collaborative?: boolean;
  description?: string;
}

export interface Playlists extends Params, Page {}

export interface UsersPlaylists extends Params, Page {}

export interface Playlist extends Params {
  fields?: string;
}

export interface PlaylistTracks extends Params, Page {
  fields?: string;
}

export interface RemovePlaylistTracks extends Params {
  snapshot_id?: string;
  tracks: [
    {
      uri: string;
      positions: string[];
    }
  ];
}

export interface ReorderPlaylistTracks extends Params {
  range_start: number;
  range_length?: number;
  insert_before: number;
  snapshot_id?: string;
}

export interface ReplacePlaylistTracks extends Params {
  uris?: string[];
}

export interface Search extends Params, Page {
  q: string;
  type: string; // album,artist,playlist,track
  include_external?: 'audio';
}

export interface RecentTracks extends Params, Cursor {}
