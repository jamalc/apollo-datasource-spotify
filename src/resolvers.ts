/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

export const resolvers = {
  Query: {
    spotify: () => ({}),
  },
  SpotifyQuery: {
    getMultipleAlbums(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetMultipleAlbums(args.params);
    },
    getAlbum(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetAlbum(args.id, args.params);
    },
    getAlbumsTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetAlbumsTracks(args.id, args.params);
    },
    getMultipleArtists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetMultipleArtists(args.params);
    },
    getArtist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetArtist(args.id);
    },
    getArtistsTopTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetArtistsTopTracks(args.id, args.params);
    },
    getArtistsRelatedArtists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetArtistsRelatedArtists(args.id);
    },
    getArtistsAlbums(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetArtistsAlbums(args.id, args.params);
    },
    getNewReleases(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetNewReleases(args.params);
    },
    getFeaturedPlaylists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetFeaturedPlaylists(args.params);
    },
    getCategories(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetCategories(args.params);
    },
    getCategory(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetCategory(args.category_id, args.params);
    },
    getCategoriesPlaylists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetCategoriesPlaylists(
        args.category_id,
        args.params
      );
    },
    getRecommendations(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetRecommendations(args.params);
    },
    getRecommendationGenres(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetRecommendationGenres();
    },
    getMultipleEpisodes(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetMultipleEpisodes(args.params);
    },
    getEpisode(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetEpisode(args.id, args.params);
    },
    checkIfUserFollowsPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckIfUserFollowsPlaylist(
        args.playlist_id,
        args.params
      );
    },
    getFollowed(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetFollowed(args.params);
    },
    checkCurrentUserFollows(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckCurrentUserFollows(args.params);
    },
    getUsersSavedAlbums(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersSavedAlbums(args.params);
    },
    checkUsersSavedAlbums(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckUsersSavedAlbums(args.params);
    },
    getUsersSavedTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersSavedTracks(args.params);
    },
    checkUsersSavedTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckUsersSavedTracks(args.params);
    },
    getUsersSavedEpisodes(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersSavedEpisodes(args.params);
    },
    checkUsersSavedEpisodes(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckUsersSavedEpisodes(args.params);
    },
    getUsersSavedShows(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersSavedShows(args.params);
    },
    checkUsersSavedShows(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CheckUsersSavedShows(args.params);
    },
    getAvailableMarkets(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetAvailableMarkets();
    },
    getUsersTopArtistsAndTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersTopArtistsAndTracks(
        args.type,
        args.params
      );
    },
    getInformationAboutTheUsersCurrentPlayback(
      _: unknown,
      args: any,
      ctx: any
    ) {
      return ctx.dataSources.spotify.GetInformationAboutTheUsersCurrentPlayback(
        args.params
      );
    },
    getUsersAvailableDevices(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersAvailableDevices();
    },
    getTheUsersCurrentlyPlayingTrack(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetTheUsersCurrentlyPlayingTrack(
        args.params
      );
    },
    getRecentlyPlayed(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetRecentlyPlayed(args.params);
    },
    getListOfCurrentUsersPlaylists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetListOfCurrentUsersPlaylists(
        args.params
      );
    },
    getListUsersPlaylists(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetListUsersPlaylists(
        args.user_id,
        args.params
      );
    },
    getPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetPlaylist(args.playlist_id, args.params);
    },
    getPlaylistsTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetPlaylistsTracks(
        args.playlist_id,
        args.params
      );
    },
    getPlaylistCover(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetPlaylistCover(args.playlist_id);
    },
    search(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.Search(args.params);
    },
    getMultipleShows(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetMultipleShows(args.params);
    },
    getShow(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetShow(args.id, args.params);
    },
    getShowsEpisodes(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetShowsEpisodes(args.id, args.params);
    },
    getSeveralTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetSeveralTracks(args.params);
    },
    getTrack(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetTrack(args.id, args.params);
    },
    getSeveralAudioFeatures(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetSeveralAudioFeatures(args.params);
    },
    getAudioFeatures(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetAudioFeatures(args.id);
    },
    getAudioAnalysis(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetAudioAnalysis(args.id);
    },
    getCurrentUsersProfile(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetCurrentUsersProfile();
    },
    getUsersProfile(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.GetUsersProfile(args.user_id);
    },
  },
  SpotifyMutation: {
    followPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.FollowPlaylist(
        args.playlist_id,
        args.params
      );
    },
    unfollowPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.UnfollowPlaylist(args.playlist_id);
    },
    followArtistsUsers(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.FollowArtistsUsers(args.params);
    },
    unfollowArtistsUsers(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.UnfollowArtistsUsers(args.params);
    },
    saveAlbumsUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SaveAlbumsUser(args.params);
    },
    removeAlbumsUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.RemoveAlbumsUser(args.params);
    },
    saveTracksUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SaveTracksUser(args.params);
    },
    removeTracksUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.RemoveTracksUser(args.params);
    },
    saveEpisodesUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SaveEpisodesUser(args.params);
    },
    removeEpisodesUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.RemoveEpisodesUser(args.params);
    },
    saveShowsUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SaveShowsUser(args.params);
    },
    removeShowsUser(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.RemoveShowsUser(args.params);
    },
    transferUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.TransferUsersPlayback(args.params);
    },
    startUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.StartUsersPlayback(args.params);
    },
    pauseUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.PauseUsersPlayback(args.params);
    },
    skipUsersPlaybackToNextTrack(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SkipUsersPlaybackToNextTrack(args.params);
    },
    skipUsersPlaybackToPreviousTrack(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SkipUsersPlaybackToPreviousTrack(
        args.params
      );
    },
    seekToPositionInCurrentlyPlayingTrack(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SeekToPositionInCurrentlyPlayingTrack(
        args.params
      );
    },
    setRepeatModeOnUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SetRepeatModeOnUsersPlayback(args.params);
    },
    setVolumeForUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.SetVolumeForUsersPlayback(args.params);
    },
    toggleShuffleForUsersPlayback(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.ToggleShuffleForUsersPlayback(args.params);
    },
    addToQueue(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.AddToQueue(args.params);
    },
    createPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.CreatePlaylist(args.user_id, args.params);
    },
    changePlaylistDetails(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.ChangePlaylistDetails(
        args.playlist_id,
        args.params
      );
    },
    addTracksToPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.AddTracksToPlaylist(
        args.playlist_id,
        args.params
      );
    },
    reorderOrReplacePlaylistsTracks(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.ReorderOrReplacePlaylistsTracks(
        args.playlist_id,
        args.params
      );
    },
    removeTracksPlaylist(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.RemoveTracksPlaylist(
        args.playlist_id,
        args.params
      );
    },
    uploadCustomPlaylistCover(_: unknown, args: any, ctx: any) {
      return ctx.dataSources.spotify.UploadCustomPlaylistCover(
        args.playlist_id
      );
    },
  },
  PagingObjectItem: {
    __resolveType(object: any) {
      const {
        type,
        album,
        track,
        episode,
        show,
        added_at,
        added_by,
        followers,
        genres,
        images,
        popularity,
        available_markets,
        copyrights,
        external_ids,
        episodes,
        played_at,
        icons,
        name,
      } = object;

      switch (type) {
        case 'album':
          return available_markets ||
            copyrights ||
            external_ids ||
            genres ||
            popularity
            ? 'AlbumObject'
            : 'SimplifiedAlbumObject';
        case 'artist':
          return followers || genres || images || popularity
            ? 'ArtistObject'
            : 'SimplifiedArtistObject';
        case 'playlist':
          return followers ? 'PlaylistObject' : 'SimplifiedPlaylistObject';
        case 'track':
          return album || external_ids || popularity
            ? 'TrackObject'
            : 'SimplifiedTrackObject';
        case 'show':
          return episodes ? 'ShowObject' : 'SimplifiedShowObject';
        case 'episode':
          return show ? 'EpisodeObject' : 'SimplifiedEpisodeObject';

        default:
          if (track && added_by) {
            return 'PlaylistTrackObject';
          }
          if (album && added_at) {
            return 'SavedAlbumObject';
          }
          if (track && added_at) {
            return 'SavedTrackObject';
          }
          if (episode && added_at) {
            return 'SavedEpisodeObject';
          }
          if (show && added_at) {
            return 'SavedShowObject';
          }
          if (track && played_at) {
            return 'PlayHistoryObject';
          }
          if (icons && name) {
            return 'CategoryObject';
          }
          return 'JSON';
      }
    },
  },
};
