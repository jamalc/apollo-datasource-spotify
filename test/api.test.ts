import { SpotifyAPI } from '../src/api';
import { token } from './token';

const api = new SpotifyAPI();
api.initialize({ context: { token } } as any);

test.todo('batchAlbum');

test.todo('batchAlbums');

test.todo('batchArtist');

test.todo('batchArtists');

test.todo('batchTrack');

test.todo('batchTracks');

test.todo('GetMultipleAlbums');

test('GetAlbum - Lift Your Skinny Fists Like Antennas to Heaven', async () => {
  const id = '2rT82YYlV9UoxBYLIezkRq';
  expect(await api.GetAlbum(id, { market: 'US' })).toMatchSnapshot();
});

test.todo('GetAlbumsTracks');

test.todo('GetMultipleArtists');
test('GetArtist - Frank Ocean', async () => {
  const id = '2h93pZq0e7k5yf4dywlkpM';
  expect(await api.GetArtist(id)).toMatchSnapshot();
});

test.todo('GetArtistsTopTracks');

test.todo('GetArtistsRelatedArtists');

test.todo('GetArtistsAlbums');

test.todo('GetNewReleases');

test.todo('GetFeaturedPlaylists');

test.todo('GetCategories');

test.todo('GetCategory');

test.todo('GetCategoriesPlaylists');

test.todo('GetRecommendations');

test.todo('GetRecommendationGenres');

test.todo('GetMultipleEpisodes');

test.todo('GetEpisode');

test.todo('FollowPlaylist');

test.todo('UnfollowPlaylist');

test.todo('CheckIfUserFollowsPlaylist');

test.todo('GetFollowed');

test.todo('FollowArtistsUsers');

test.todo('UnfollowArtistsUsers');

test.todo('CheckCurrentUserFollows');

test.todo('GetUsersSavedAlbums');

test.todo('SaveAlbumsUser');

test.todo('RemoveAlbumsUser');

test.todo('CheckUsersSavedAlbums');

test.todo('GetUsersSavedTracks');

test.todo('SaveTracksUser');

test.todo('RemoveTracksUser');

test.todo('CheckUsersSavedTracks');

test.todo('GetUsersSavedEpisodes');

test.todo('SaveEpisodesUser');

test.todo('RemoveEpisodesUser');

test.todo('CheckUsersSavedEpisodes');

test.todo('GetUsersSavedShows');

test.todo('SaveShowsUser');

test.todo('RemoveShowsUser');

test.todo('CheckUsersSavedShows');

test.todo('GetAvailableMarkets');

test.todo('GetUsersTopArtistsAndTracks');

test.todo('GetInformationAboutTheUsersCurrentPlayback');

test.todo('TransferUsersPlayback');

test.todo('GetUsersAvailableDevices');

test.todo('GetTheUsersCurrentlyPlayingTrack');

test.todo('StartUsersPlayback');

test.todo('PauseUsersPlayback');

test.todo('SkipUsersPlaybackToNextTrack');

test.todo('SkipUsersPlaybackToPreviousTrack');

test.todo('SeekToPositionInCurrentlyPlayingTrack');

test.todo('SetRepeatModeOnUsersPlayback');

test.todo('SetVolumeForUsersPlayback');

test.todo('ToggleShuffleForUsersPlayback');

test.todo('GetRecentlyPlayed');

test.todo('AddToQueue');

test.todo('GetListOfCurrentUsersPlaylists');

test.todo('GetListUsersPlaylists');

test.todo('CreatePlaylist');

test.todo('GetPlaylist');

test.todo('ChangePlaylistDetails');

test.todo('GetPlaylistsTracks');

test.todo('AddTracksToPlaylist');

test.todo('ReorderOrReplacePlaylistsTracks');

test.todo('RemoveTracksPlaylist');

test.todo('GetPlaylistCover');

test.todo('UploadCustomPlaylistCover');

test.todo('Search');

test.todo('GetMultipleShows');

test.todo('GetShow');

test.todo('GetShowsEpisodes');

test.todo('GetSeveralTracks');
test('GetTrack - Sketch for Summer', async () => {
  const id = '2oeLt7N1amHZiev27FvKE2?si=a03757f17a05462b';
  expect(await api.GetTrack(id, { market: 'US' })).toMatchSnapshot();
});

test.todo('GetSeveralAudioFeatures');

test.todo('GetAudioFeatures');

test.todo('GetAudioAnalysis');

test.todo('GetCurrentUsersProfile');

test.todo('GetUsersProfile');
