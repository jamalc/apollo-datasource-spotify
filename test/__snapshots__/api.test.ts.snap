// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAlbum - Lift Your Skinny Fists Like Antennas to Heaven 1`] = `
Object {
  "album_type": "album",
  "artists": Array [
    Object {
      "external_urls": Object {
        "spotify": "https://open.spotify.com/artist/4svpOyfmQKuWpHLjgy4cdK",
      },
      "href": "https://api.spotify.com/v1/artists/4svpOyfmQKuWpHLjgy4cdK",
      "id": "4svpOyfmQKuWpHLjgy4cdK",
      "name": "Godspeed You! Black Emperor",
      "type": "artist",
      "uri": "spotify:artist:4svpOyfmQKuWpHLjgy4cdK",
    },
  ],
  "copyrights": Array [
    Object {
      "text": "(C) 2000 kranky",
      "type": "C",
    },
  ],
  "external_ids": Object {
    "upc": "887396825833",
  },
  "external_urls": Object {
    "spotify": "https://open.spotify.com/album/2rT82YYlV9UoxBYLIezkRq",
  },
  "genres": Array [],
  "href": "https://api.spotify.com/v1/albums/2rT82YYlV9UoxBYLIezkRq",
  "id": "2rT82YYlV9UoxBYLIezkRq",
  "images": Array [
    Object {
      "height": 640,
      "url": "https://i.scdn.co/image/ab67616d0000b2730791c8a2dc44008c52b4883a",
      "width": 640,
    },
    Object {
      "height": 300,
      "url": "https://i.scdn.co/image/ab67616d00001e020791c8a2dc44008c52b4883a",
      "width": 300,
    },
    Object {
      "height": 64,
      "url": "https://i.scdn.co/image/ab67616d000048510791c8a2dc44008c52b4883a",
      "width": 64,
    },
  ],
  "label": "kranky",
  "name": "Lift Your Skinny Fists Like Antennas to Heaven",
  "popularity": 45,
  "release_date": "2000-01-01",
  "release_date_precision": "day",
  "total_tracks": 4,
  "tracks": Object {
    "href": "https://api.spotify.com/v1/albums/2rT82YYlV9UoxBYLIezkRq/tracks?offset=0&limit=50&market=US",
    "items": Array [
      Object {
        "artists": Array [
          Object {
            "external_urls": Object {
              "spotify": "https://open.spotify.com/artist/4svpOyfmQKuWpHLjgy4cdK",
            },
            "href": "https://api.spotify.com/v1/artists/4svpOyfmQKuWpHLjgy4cdK",
            "id": "4svpOyfmQKuWpHLjgy4cdK",
            "name": "Godspeed You! Black Emperor",
            "type": "artist",
            "uri": "spotify:artist:4svpOyfmQKuWpHLjgy4cdK",
          },
        ],
        "disc_number": 1,
        "duration_ms": 1352413,
        "explicit": false,
        "external_urls": Object {
          "spotify": "https://open.spotify.com/track/1HfJV18PHF2UQqh4TuySBJ",
        },
        "href": "https://api.spotify.com/v1/tracks/1HfJV18PHF2UQqh4TuySBJ",
        "id": "1HfJV18PHF2UQqh4TuySBJ",
        "is_local": false,
        "is_playable": true,
        "name": "Storm",
        "preview_url": "https://p.scdn.co/mp3-preview/a71a88ded16564d37e985c28e89584d880b6830a?cid=00a39221303745a383e9087973d97693",
        "track_number": 1,
        "type": "track",
        "uri": "spotify:track:1HfJV18PHF2UQqh4TuySBJ",
      },
      Object {
        "artists": Array [
          Object {
            "external_urls": Object {
              "spotify": "https://open.spotify.com/artist/4svpOyfmQKuWpHLjgy4cdK",
            },
            "href": "https://api.spotify.com/v1/artists/4svpOyfmQKuWpHLjgy4cdK",
            "id": "4svpOyfmQKuWpHLjgy4cdK",
            "name": "Godspeed You! Black Emperor",
            "type": "artist",
            "uri": "spotify:artist:4svpOyfmQKuWpHLjgy4cdK",
          },
        ],
        "disc_number": 1,
        "duration_ms": 1355946,
        "explicit": false,
        "external_urls": Object {
          "spotify": "https://open.spotify.com/track/1mVrFcCEpYcF6ISXGsyRk5",
        },
        "href": "https://api.spotify.com/v1/tracks/1mVrFcCEpYcF6ISXGsyRk5",
        "id": "1mVrFcCEpYcF6ISXGsyRk5",
        "is_local": false,
        "is_playable": true,
        "name": "Static",
        "preview_url": "https://p.scdn.co/mp3-preview/6d1257efe118ed1dfa173d1cd37a5123024146ef?cid=00a39221303745a383e9087973d97693",
        "track_number": 2,
        "type": "track",
        "uri": "spotify:track:1mVrFcCEpYcF6ISXGsyRk5",
      },
      Object {
        "artists": Array [
          Object {
            "external_urls": Object {
              "spotify": "https://open.spotify.com/artist/4svpOyfmQKuWpHLjgy4cdK",
            },
            "href": "https://api.spotify.com/v1/artists/4svpOyfmQKuWpHLjgy4cdK",
            "id": "4svpOyfmQKuWpHLjgy4cdK",
            "name": "Godspeed You! Black Emperor",
            "type": "artist",
            "uri": "spotify:artist:4svpOyfmQKuWpHLjgy4cdK",
          },
        ],
        "disc_number": 2,
        "duration_ms": 1397746,
        "explicit": false,
        "external_urls": Object {
          "spotify": "https://open.spotify.com/track/05JqOBN6XW4eFUVQlgR0I3",
        },
        "href": "https://api.spotify.com/v1/tracks/05JqOBN6XW4eFUVQlgR0I3",
        "id": "05JqOBN6XW4eFUVQlgR0I3",
        "is_local": false,
        "is_playable": true,
        "name": "Sleep",
        "preview_url": "https://p.scdn.co/mp3-preview/9be5a84b2797bb064223a8336b0103a85d48cecc?cid=00a39221303745a383e9087973d97693",
        "track_number": 1,
        "type": "track",
        "uri": "spotify:track:05JqOBN6XW4eFUVQlgR0I3",
      },
      Object {
        "artists": Array [
          Object {
            "external_urls": Object {
              "spotify": "https://open.spotify.com/artist/4svpOyfmQKuWpHLjgy4cdK",
            },
            "href": "https://api.spotify.com/v1/artists/4svpOyfmQKuWpHLjgy4cdK",
            "id": "4svpOyfmQKuWpHLjgy4cdK",
            "name": "Godspeed You! Black Emperor",
            "type": "artist",
            "uri": "spotify:artist:4svpOyfmQKuWpHLjgy4cdK",
          },
        ],
        "disc_number": 2,
        "duration_ms": 1137586,
        "explicit": false,
        "external_urls": Object {
          "spotify": "https://open.spotify.com/track/31pLszBSGG3ToeXcPIgyXF",
        },
        "href": "https://api.spotify.com/v1/tracks/31pLszBSGG3ToeXcPIgyXF",
        "id": "31pLszBSGG3ToeXcPIgyXF",
        "is_local": false,
        "is_playable": true,
        "name": "Like Antennas To Heavenâ€¦",
        "preview_url": "https://p.scdn.co/mp3-preview/d22a60819d0ae428e9bc9bb51ca5668f2f8e3fe2?cid=00a39221303745a383e9087973d97693",
        "track_number": 2,
        "type": "track",
        "uri": "spotify:track:31pLszBSGG3ToeXcPIgyXF",
      },
    ],
    "limit": 50,
    "next": null,
    "offset": 0,
    "previous": null,
    "total": 4,
  },
  "type": "album",
  "uri": "spotify:album:2rT82YYlV9UoxBYLIezkRq",
}
`;

exports[`GetArtist - Frank Ocean 1`] = `
Object {
  "external_urls": Object {
    "spotify": "https://open.spotify.com/artist/2h93pZq0e7k5yf4dywlkpM",
  },
  "followers": Object {
    "href": null,
    "total": 7728869,
  },
  "genres": Array [
    "alternative r&b",
    "hip hop",
    "lgbtq+ hip hop",
    "neo soul",
    "pop",
  ],
  "href": "https://api.spotify.com/v1/artists/2h93pZq0e7k5yf4dywlkpM",
  "id": "2h93pZq0e7k5yf4dywlkpM",
  "images": Array [
    Object {
      "height": 640,
      "url": "https://i.scdn.co/image/7db34c8aace6feb91f38601bb75e6b3301b4657a",
      "width": 640,
    },
    Object {
      "height": 320,
      "url": "https://i.scdn.co/image/0cc22250c0b18183e5c62f240a5756ec5226dee1",
      "width": 320,
    },
    Object {
      "height": 160,
      "url": "https://i.scdn.co/image/e42028e7fe5a7ba790b5b4e7b096b06bf3319443",
      "width": 160,
    },
  ],
  "name": "Frank Ocean",
  "popularity": 87,
  "type": "artist",
  "uri": "spotify:artist:2h93pZq0e7k5yf4dywlkpM",
}
`;

exports[`GetTrack - Sketch for Summer 1`] = `
Object {
  "album": Object {
    "album_type": "album",
    "artists": Array [
      Object {
        "external_urls": Object {
          "spotify": "https://open.spotify.com/artist/7fh7bwX9qV60tLxxsp9bTe",
        },
        "href": "https://api.spotify.com/v1/artists/7fh7bwX9qV60tLxxsp9bTe",
        "id": "7fh7bwX9qV60tLxxsp9bTe",
        "name": "The Durutti Column",
        "type": "artist",
        "uri": "spotify:artist:7fh7bwX9qV60tLxxsp9bTe",
      },
    ],
    "external_urls": Object {
      "spotify": "https://open.spotify.com/album/6ANCHHVpcvjCqQFUDUvDOW",
    },
    "href": "https://api.spotify.com/v1/albums/6ANCHHVpcvjCqQFUDUvDOW",
    "id": "6ANCHHVpcvjCqQFUDUvDOW",
    "images": Array [
      Object {
        "height": 640,
        "url": "https://i.scdn.co/image/ab67616d0000b2734be74b9209d99dc3a8c682e8",
        "width": 640,
      },
      Object {
        "height": 300,
        "url": "https://i.scdn.co/image/ab67616d00001e024be74b9209d99dc3a8c682e8",
        "width": 300,
      },
      Object {
        "height": 64,
        "url": "https://i.scdn.co/image/ab67616d000048514be74b9209d99dc3a8c682e8",
        "width": 64,
      },
    ],
    "name": "The Return of The Durutti Column",
    "release_date": "1979-01-01",
    "release_date_precision": "day",
    "total_tracks": 15,
    "type": "album",
    "uri": "spotify:album:6ANCHHVpcvjCqQFUDUvDOW",
  },
  "artists": Array [
    Object {
      "external_urls": Object {
        "spotify": "https://open.spotify.com/artist/7fh7bwX9qV60tLxxsp9bTe",
      },
      "href": "https://api.spotify.com/v1/artists/7fh7bwX9qV60tLxxsp9bTe",
      "id": "7fh7bwX9qV60tLxxsp9bTe",
      "name": "The Durutti Column",
      "type": "artist",
      "uri": "spotify:artist:7fh7bwX9qV60tLxxsp9bTe",
    },
  ],
  "disc_number": 1,
  "duration_ms": 178280,
  "explicit": false,
  "external_ids": Object {
    "isrc": "GBAAP0200153",
  },
  "external_urls": Object {
    "spotify": "https://open.spotify.com/track/2oeLt7N1amHZiev27FvKE2",
  },
  "href": "https://api.spotify.com/v1/tracks/2oeLt7N1amHZiev27FvKE2",
  "id": "2oeLt7N1amHZiev27FvKE2",
  "is_local": false,
  "is_playable": true,
  "name": "Sketch for Summer",
  "popularity": 54,
  "preview_url": "https://p.scdn.co/mp3-preview/dd1c62b4a56524bff11306f5f40db156230bdc2c?cid=00a39221303745a383e9087973d97693",
  "track_number": 1,
  "type": "track",
  "uri": "spotify:track:2oeLt7N1amHZiev27FvKE2",
}
`;
